### PR TITLE
Add Support for Razer Atheris

### LINKED ISSUE #595: Razer Atheris support
I want to buy this mouse. Is it possible to configure its buttons with openrazer?
Will my macroses work with bluetooth connection(as I know this mouse don't have internal memory)?

### PR DESCRIPTION
**Pull request by nightsky30**
Add Support for Razer Atheris

- [x] Serial
- [x] Firmware
- [x] DPI
- [x] Polling Rate
- [x] Battery Status

Notes:
* Requires an insanely long wait time for return values compared to other devices.
* Despite the long wait times, battery status only returns correct values when the frontend is NOT open. If the frontend IS displayed, and the user inserts the USB receiver, the battery status displays an incorrect value. If you close the frontend and then reinsert the USB receiver or just wait for another status to display, the battery will report the proper value once again.
* I have tried increasing the wait times far in excess of the current value (like ~800000, 900000, 1200000), and it did not have any further impact on the battery status reported when the frontend was open. See [comment](https://github.com/openrazer/openrazer/pull/1203#issuecomment-660629605).
* Bluetooth mode not supported via openrazer, but basic pairing/usage seems to work. See [comment](https://github.com/openrazer/openrazer/pull/1203#issuecomment-663992279).
* If the user plugs in the 2.4GHz receiver, but the mouse is off or set to Bluetooth mode, the user will have problems.
* The mouse (not the receiver) appears to be storing the DPI and polling rate as no values are returned if the receiver is plugged in and the mouse is powered off.
* Settings persist even if the 2xAA batteries are removed.
* The 2.4GHz receiver appears to own the PID 0x0062, while the mouse itself has the PID 0x0061 while powered in Bluetooth mode.
* There are issues with freezing or jittery [movement](https://github.com/openrazer/openrazer/issues/595#issuecomment-506974403). This appears to be an issue with [firmware](https://support.razer.com/gaming-mice-and-mats/razer-atheris/), not Linux or openrazer.
* Razer claimed to have fixed the freezing with the latest [firmware](https://support.razer.com/gaming-mice-and-mats/razer-atheris/), but the newest version available at this time (v1.3) still seems to have the issues.

**EDIT:**
* Latest firmware version available appears to be v1.5 [here](https://mysupport.razer.com/app/answers/detail/a_id/479/kw/firmware).

Fixes #595

### CODE REVIEW

**Comment by nightsky30** [comment]
Still seeing some instability while retesting on bare metal with other devices vs VM with singular device...

**Comment by nightsky30** [comment]
This is what I am seeing when plugging in the mouse receiver with RazerGenie open and the battery reports a wrong value.

It reports fine when plugging in the receiver and RazerGenie is closed. The error seems to indicate some kind of timing issue as it is consistent with other errors with reading values (DPI, poll rate, serial, firmware) that have become less common as wait times were increased, and it matches what was reported by bno1 with the Basilisk X Hyperspeed https://github.com/openrazer/openrazer/pull/1055#issuecomment-622203827

```
[  119.622973] usb 1-9: new full-speed USB device number 13 using xhci_hcd
[  119.773938] usb 1-9: New USB device found, idVendor=1532, idProduct=0062, bcdDevice= 2.00
[  119.773944] usb 1-9: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[  119.773947] usb 1-9: Product: Razer Atheris - Mobile Gaming Mouse
[  119.773950] usb 1-9: Manufacturer: Razer
[  119.779878] input: Razer Razer Atheris - Mobile Gaming Mouse as /devices/pci0000:00/0000:00:14.0/usb1/1-9/1-9:1.0/0003:1532:0062.0017/input/input60
[  119.780522] razermouse 0003:1532:0062.0017: input,hidraw0: USB HID v1.11 Mouse [Razer Razer Atheris - Mobile Gaming Mouse] on usb-0000:00:14.0-9/input0
[  119.783481] input: Razer Razer Atheris - Mobile Gaming Mouse as /devices/pci0000:00/0000:00:14.0/usb1/1-9/1-9:1.1/0003:1532:0062.0018/input/input61
[  119.843383] razermouse 0003:1532:0062.0018: input,hidraw3: USB HID v1.11 Keyboard [Razer Razer Atheris - Mobile Gaming Mouse] on usb-0000:00:14.0-9/input1
[  119.845161] input: Razer Razer Atheris - Mobile Gaming Mouse as /devices/pci0000:00/0000:00:14.0/usb1/1-9/1-9:1.2/0003:1532:0062.0019/input/input62
[  119.903570] razermouse 0003:1532:0062.0019: input,hidraw5: USB HID v1.11 Keyboard [Razer Razer Atheris - Mobile Gaming Mouse] on usb-0000:00:14.0-9/input2
[  122.941403] razermouse: Response doesn't match request. Start Marker: 02 id: ff Num Params: 07 Reserved: 04 Command: 85 Params: 00070807080000000000000000000000 .
```

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 1]
```diff
@@ -2067,3 +2067,44 @@ def _resume_device(self):
...
+        "top_img": "https://assets.razerzone.com/eeimages/support/products/1234/1234_atheris.png",
+    }
+
+    DPI_MAX = 7200
+
+    def _resume_device(self):
+        self.logger.debug("Atheris doesn't have suspend/resume")
+
+    def _suspend_device(self):
+        self.logger.debug("Atheris doesn't have suspend/resume")
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: I believe you can leave out these methods if they're not required?

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 2]
```diff
@@ -2067,3 +2067,44 @@ def _resume_device(self):
...
+class RazerAtheris(__RazerDevice):
+    """
+    Class for the Razer Atheris
+    """
+    USB_VID = 0x1532
+    USB_PID = 0x0062
+    EVENT_FILE_REGEX = re.compile(r'.*Razer_Razer_Atheris_-_Mobile_Gaming_Mouse-if0(1|2)-event-kbd')
+    METHODS = ['get_device_type_mouse', 'max_dpi', 'get_dpi_xy', 'set_dpi_xy', 'get_poll_rate', 'set_poll_rate',
+               # Battery
+               'get_battery', 'set_idle_time', 'set_low_battery_threshold']
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: no `is_charging`?

File: driver/razermouse_driver.h [hunk 3]
```diff
@@ -72,6 +73,9 @@
 #define RAZER_NEW_MOUSE_RECEIVER_WAIT_MIN_US 31000
 #define RAZER_NEW_MOUSE_RECEIVER_WAIT_MAX_US 31100
 
+#define RAZER_ATHERIS_RECEIVER_WAIT_MIN_US 400000
+#define RAZER_ATHERIS_RECEIVER_WAIT_MAX_US 400100
```
Comment (z3ntu) [inline driver/razermouse_driver.h:79]: wtf :facepalm: 

(you can just 'resolve' this, just wanted to add a comment)

**Review by z3ntu — CHANGES_REQUESTED** [review]
generally looks fine, just a few things

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 2 above)
Comment (nightsky30) [reply daemon/openrazer_daemon/hardware/mouse.py]: It uses 2xAA batteries.

File: driver/razermouse_driver.h (hunk 3 above)
Comment (nightsky30) [reply driver/razermouse_driver.h:79]: YES :rofl:

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 1 above)
Comment (nightsky30) [reply daemon/openrazer_daemon/hardware/mouse.py]: Ok, done :)

**Comment by Samyak2** [comment]
I have tried this driver with my Razer Atheris using the build instructions from [here](https://github.com/openrazer/openrazer/wiki/Building-a-package). I had the same issue as #1205 but fixed with the help of the replies there. Also noticed #877 but couldn't fix it (other than re-connecting).

 - Works on 2.4GHz, but setting values takes a few seconds to take effect. I could only see (in RazerGenie) DPI and polling rate settings, there was no battery status.
 - On bluetooth, it only shows polling rate correctly. DPI and Firmware version are 0.

Thank you for this @nightsky30.

**Comment by nightsky30** [comment]
Thanks for testing as well @Samyak2 :smile: 
- Battery status is only displayed through desktop notifications at the moment. Future releases of the front-ends should include some type of battery status.
- I'd tried Bluetooth, but didn't seem to get any read/write setting indications when the mouse was paired. I tried again to use Bluetooth, and I can't pair the mouse now because blueman-adapters is crashing saying no adapters found...

**Comment by Samyak2** [comment]
- Oh okay. Is there a frontend which shows the battery status? Though I can see the battery percentage through bluetooth, it would be nice to have it in the frontend too.
- It works perfectly on bluetooth for me, connects instantly once paired. RazerGenie also recognises it but I cannot change anything.

**Comment by nightsky30** [comment]
No, no front-end shows the battery status now. Polychromatic will definitely have it in a future release. I'm guessing maybe RazerGenie will want to include it as well.

Oh wonderful on the Bluetooth news. I'll try and get mine working. Interesting that settings appear, but cannot be changed.

**Comment by nightsky30** [comment]
Ok, so...

Apparently using the USB pass-through to allow your host's Bluetooth to be discovered and used in Virtualbox by a Windows guest does something that persists after reboots and even full shutdowns of the host system. My Bluetooth service appeared to be functioning fine, but bluetoothctl continued to report no default controller found **UNTIL** I shutdown and **PULLED MY POWER CABLE** for 30 seconds. Bluetooth is working fine again.

I tried the mouse in Bluetooth mode, and again the mouse functions, but openrazer doesn't detect the device or expose anything to the front-ends. I found while in BT mode the mouse actually has a different PID of x0061 vs the 2.4GHz receiver mode which has a PID of x0062. 

After realizing there was the different PID, I actually tried adding a second entry for the Bluetooth mode to the mouse driver. But because the device and input appear to be handled differently in BT mode, it isn't working.

For reference:
```
$ udevadm info -a /dev/input/mouse2

Udevadm info starts with the device specified by the devpath and then
walks up the chain of parent devices. It prints for every device
found, all possible attributes in the udev rules key format.
A rule to match, can be composed by the attributes of the device
and the attributes from one single parent device.

  looking at device '/devices/virtual/misc/uhid/0005:1532:0061.0017/input/input84/mouse2':
    KERNEL=="mouse2"
    SUBSYSTEM=="input"
    DRIVER==""

  looking at parent device '/devices/virtual/misc/uhid/0005:1532:0061.0017/input/input84':
    KERNELS=="input84"
    SUBSYSTEMS=="input"
    DRIVERS==""
    ATTRS{phys}==":..."
    ATTRS{name}=="Razer Atheris Mouse"
    ATTRS{uniq}==":..."
    ATTRS{properties}=="0"

  looking at parent device '/devices/virtual/misc/uhid/0005:1532:0061.0017':
    KERNELS=="0005:1532:0061.0017"
    SUBSYSTEMS=="hid"
    DRIVERS=="hid-generic"
    ATTRS{country}=="00"

  looking at parent device '/devices/virtual/misc/uhid':
    KERNELS=="uhid"
    SUBSYSTEMS=="misc"
    DRIVERS==""
```

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 2 above)
Comment (z3ntu) [reply daemon/openrazer_daemon/hardware/mouse.py]: please add is_charging like done in the Basilisk X Hyperspeed (incl. hardcoded "0" in driver)

**Review by z3ntu — CHANGES_REQUESTED** [review]
other than the is_charging thing LGTM

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 2 above)
Comment (nightsky30) [reply daemon/openrazer_daemon/hardware/mouse.py]: Ok, it is added :)

**Review by z3ntu — APPROVED** [review]
LGTM, thanks as always!

**Pull request merged**